politeiawww/addresswatcher.go

```diff
@@ -0,0 +1,314 @@
+					msg.EventId, m.Address, m.TxHash)
+
+				// Check payment history for address
+				payment, err := p.cmsDB.PaymentsByAddress(m.Address)
+				if err != nil {
+					log.Errorf("error retreiving payments information from db %v", err)
+				}
+				paid := p.checkPayments(payment)
+				if paid {
+					p.removeWatchAddress(payment.Address)
+				}
+			case *pstypes.TxList:
+				log.Debugf("Message (%s): TxList(len=%d)", msg.EventId, len(*m))
+			default:
+				log.Debugf("Message of type %v unhandled. %v", msg.EventId, m)
+			}
+		}
+	}()
+	return nil
+}
+
+func (p *politeiawww) restartAddressesWatching() error {
+	approvedInvoices, err := p.cmsDB.InvoicesByStatus(int(cms.InvoiceStatusApproved))
+	if err != nil {
+		return err
+	}
+	for _, invoice := range approvedInvoices {
+		_, err := p.cmsDB.PaymentsByAddress(invoice.PaymentAddress)
+		if err != nil {
+			if err == database.ErrInvoiceNotFound {
+				payout, err := p.calculatePayout(invoice)
+				if err != nil {
+					return err
+				}
+				// Start listening the first day of the next month of invoice.
+				listenStartDate := time.Date(int(invoice.Year),
+					time.Month(invoice.Month+1), 0, 0, 0, 0, 0, time.UTC)
+				invoice.Payments = database.Payments{
+					Address:      invoice.PaymentAddress,
+					TimeStarted:  listenStartDate.Unix(),
+					Status:       cms.PaymentStatusWatching,
+					AmountNeeded: int64(payout.DCRTotal),
+				}
+				fmt.Println(listenStartDate)
+				err = p.cmsDB.UpdateInvoice(&invoice)
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+	unpaidPayments, err := p.cmsDB.PaymentsByStatus(uint(cms.PaymentStatusWatching))
+	if err != nil {
+		return err
+	}
+	for _, payments := range unpaidPayments {
+		paid := p.checkPayments(&payments)
+		if !paid {
+			p.addWatchAddress(payments.Address)
+		}
+	}
+	return nil
+}
+
+func strInSlice(sl []string, str string) (bool, int) {
+	for i, s := range sl {
+		if s == str {
+			return true, i
+		}
+	}
+	return false, -1
+}
+
+func (p *politeiawww) checkPayments(payment *database.Payments) bool {
+	// Get all txs since start time of watcher
+	txs, err := util.FetchTxsForAddressNotBefore(payment.Address, payment.TimeStarted)
+	if err != nil {
+		// XXX Some sort of 'recheck' or notice that it should do it again?
+		log.Errorf("error FetchTxsForAddressNotBefore for %s", payment.Address)
+	}
+	if len(txs) == len(payment.TxIDs) {
+		// Same number of txids found, so nothing to update.
+		return false
+	}
+
+	txIDs := ""
+	// Calculate amount received
+	amountReceived := dcrutil.Amount(0)
+	for i, tx := range txs {
+		amountReceived += dcrutil.Amount(tx.Amount)
+		if i == 0 {
+			txIDs = tx.TxID
+		} else {
+			txIDs += ", " + tx.TxID
+		}
+	}
+	payment.TxIDs = txIDs
+
+	if int64(amountReceived) == payment.AmountReceived {
+		// Amount received still the same so nothing to update.
+		return false
+	}
+
+	if int64(amountReceived) >= payment.AmountNeeded {
+		payment.Status = cms.PaymentStatusPaid
```

Comment:
What if the person is using an address that they also use for other stuff?  Do we need to check if the payment is coming from the network treasury?

Reply:
Yah was going to add that, though wasn't sure how to handle it for testing.  Maybe only enforce that rule for Mainnet?

Reply:
Yeah, only enforce on mainnet.